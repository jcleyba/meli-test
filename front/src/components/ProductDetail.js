import React, { Component } from 'react';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import ProgressiveImage from 'react-progressive-image'
import Helmet from "react-helmet";

import { fetchItemById } from '../actions/productsActions';
import Spinner from './common/Spinner';
import Breadcrumbs from './common/Breadcrumbs';

import './ProductDetails.css';

export class ProductDetail extends Component {

  componentWillMount = () => {
    const id = this.props.match.params.id;
    this.props.fetchItemById(id);
  };

  renderDescription = (description) => {
    if (!description) return null;

    if (description.text.length) {
      return <div className="item-description"
                  dangerouslySetInnerHTML={{__html: description.text}}></div>
    }

    return <div className="item-description">{description.plain_text}</div>
  };

  renderImage = (product) => {
    const {title, thumbnail} = product;
    const picURI = thumbnail && thumbnail.replace('I', 'B');

    // Using progressive image for better UX on loading of images.
    return (
      <div className="item-picture">
        <ProgressiveImage src={picURI} placeholder={thumbnail}>
          {(src) => <img src={src} alt={title} title={title} className="product-pic"/>}
        </ProgressiveImage>
      </div>
    );
  };

  renderBuyButton = () => {
    return <button className="buy-btn">Comprar</button>

  };

  renderStatus = (product) => {
    const {condition} = product;
    const used = condition === 'used' ? 'Usado' : 'Nuevo';
    return (
      <div className="item-condition">
        {used} - {`${product.sold_quantity} vendidos`}
      </div>
    );
  };

  renderPrice = (product) => {
    const price = product.price && product.price.amount;
    return <div className="item-price">{`$ ${price}`}</div>
  };

  renderMetaTitle = () => {
    // Change Meta title for SEO
    return <Helmet title={`${this.props.product.title} - Mercado Libre Argentina`}/>
  };

  render() {

    const {isFetching, product, filters} = this.props;
    const pathFromRoot = filters.length && filters[0].values[0].path_from_root || [];

    if (isFetching) {
      return <Spinner/>
    }
    return (
      <div className="product-wrapper">
        {this.renderMetaTitle()}
        <Breadcrumbs pathFromRoot={pathFromRoot}/>
        <div className="product-detail">
          <div className="left">
            {this.renderImage(product)}
            <h2>Descripci√≥n del producto</h2>
            {this.renderDescription(product.description)}
          </div>
          <div className="right">
            {this.renderStatus(product)}
            <h1 className="item-title">{product.title}</h1>
            {this.renderPrice(product)}
            {this.renderBuyButton()}
          </div>
        </div>
      </div>
    )
  }
}

const mapStateToProps = ({productsReducer, searchReducer}) => {
  const {isFetching, product, error} = productsReducer;
  const {filters} = searchReducer;

  return {isFetching, product, error, filters};
};

const componentWithRouter = withRouter(ProductDetail);

export default connect(mapStateToProps, {fetchItemById})(componentWithRouter);
