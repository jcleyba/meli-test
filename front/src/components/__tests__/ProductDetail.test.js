import React from 'react';
import { shallow, mount } from 'enzyme';
import 'jest-enzyme';

import { ProductDetail } from '../ProductDetail';

describe('ProductDetail Component', () => {

  let wrapper;
  let mounted;

  const props = {
    match: {
      params: {
        id: 1
      }
    },
    product: {},
    isFetching: false,
    fetchItemById: jest.fn(),
    filters: []
  };


  beforeEach(() => {
    wrapper = shallow(<ProductDetail {...props}/>);
    mounted = mount(<ProductDetail {...props}/>);
  });

  afterEach(() => {
    wrapper = null;
    mounted = null;
  });

  describe('ProductDetail', () => {

    it('matches current snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should have isFetching prop', () => {
      expect(mounted).toHaveProp('isFetching');
    });

    it('should have product prop', () => {
      expect(mounted).toHaveProp('product');
    });

    it('should have a product-detail className present', () => {
      expect(mounted.find('.product-detail')).toBePresent();
    });

    it('should have a left section present', () => {
      expect(mounted.find('.left')).toBePresent();
    });

    it('should have a right section present', () => {
      expect(mounted.find('.right')).toBePresent();
    });

    it('should contain a product image', () => {
      expect(mounted.find('.product-pic')).toBePresent();
    });

    it('should contain a product price', () => {
      expect(mounted.find('.item-price')).toBePresent();
    });
  })

})
;
