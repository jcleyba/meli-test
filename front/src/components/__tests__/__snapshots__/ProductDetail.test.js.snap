// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetail Component ProductDetail matches current snapshot 1`] = `"<div class=\\"product-wrapper\\"><ol class=\\"breadcrumbs\\"></ol><div class=\\"product-detail\\"><div class=\\"left\\"><div class=\\"item-picture\\"><img class=\\"product-pic\\"/></div><h2>Descripci√≥n del producto</h2></div><div class=\\"right\\"><div class=\\"item-condition\\">Nuevo - undefined vendidos</div><h1 class=\\"item-title\\"></h1><div class=\\"item-price\\">$ undefined</div><button class=\\"buy-btn\\">Comprar</button></div></div></div>"`;
